Introduce PlaceEncoding machinery

dario23 opened the pull request:
This will be used for encoding array accesses. This PR includes just the machinery, no actual array encoding, but makes sure to keep all code working while using the new type.

The reasoning for the `PlaceEncoding` type is
 - we encode array code using an abstract predicate like `Array$4$i32(self: Ref)` and specify behaviour using an abstract pure function like `Array$4$i32$lookup_pure(self: Ref, index: Int): Int`. This way we don't always have a `vir::Expr` to return like for field accesses.
 - We need to encode array accesses differently for pure functions/procedures, so the `MirEncoder` returns a `PlaceEncoding::ArrayAccess` for them, and `PureFunctionEncoder` and `ProcedureEncoder` handle them accordingly
 - `PlaceEncoding::Field` and `PlaceEncoding::Variant` exist just like for `vir::Expr` with the same intent. As MIR handles deref, array indexing, field access all as a list of projections, we need to be able to mix them, so we duplicate them in `PlaceEncoding` and map them back to the respective `vir::Expr` variants when we translate the ArrayAccess.

dario23 committed:
Wrap encode_place from ProcedureEncoder

dario23 committed:
Add PlaceEncoding and postprocessing into vir::Stmt

Doesn't do any array encoding yet, but this should work just as before
on code not using any arrays, just through the PlaceEncoding mechanism
now.

dario23 committed:
Try to handle FieldAccess and Variant in PlaceEncoding::try_into_expr

fpoli commented on prusti-viper/src/encoder/mir_encoder/mod.rs:
```diff
@@ -20,13 +20,93 @@ use rustc_middle::{mir, ty};
+            PlaceEncoding::Expr(e) => write!(f, "{}", e),
+            PlaceEncoding::FieldAccess { base, field } => write!(f, "{}.{}", base, field),
+            PlaceEncoding::ArrayAccess { base, index, .. } => write!(f, "{}[{}]", base, index),
+            PlaceEncoding::Variant { base, field } => write!(f, "{}[{}]", base, field),
```
What about `"({} as {})"` to distinguish `Variant` from `ArrayAccess`?

dario23 replied:
so somewhat like the mir printing does? this is from the encoding of `prusti-tests/tests/verify/pass/enums/basic.rs` (file `log/vir_program_before_foldunfold/basic.rs.vir`, line 386 here):
```
    // [mir] _6 = &'_ mut ((_2 as V2).0: u32)
    _6 := builtin$havoc_ref()
    inhale acc(_6.val_ref, write)
    _6.val_ref := mut borrow _2[enum_V2].f$0 // L0
```

so i'd probably make a change to make the last line `_6.val_ref := mut borrow (_2 as enum_V2).f$0` instead? that log/ stuff is never read back anywhere, right? so i'd only have to change what is printed, no parser or anything else?

i'd do that in a separate PR probably, this one feels heavy already, and i quite like the "small PRs, easier/quicker reviews, faster small-step progress" approach.

fpoli replied:
I didn't notice that `field` is a string and not a number, so it should be difficult to confuse the two variants. It's fine as you did.

fpoli commented on prusti-viper/src/encoder/mir_encoder/mod.rs:
```diff
@@ -20,13 +20,93 @@ use rustc_middle::{mir, ty};
 
+/// Result of encoding a place
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum PlaceEncoding {
```
Could you move `PlaceEncoding` and its `impl` to a new `mir_encoder/place_encoding.rs` file?

fpoli commented on prusti-viper/src/encoder/mir_encoder/mod.rs:
```diff
@@ -205,15 +285,16 @@ pub trait PlaceEncoder<'v, 'tcx: 'v> {
+        encoded_base: PlaceEncoding,
         base_ty: ty::Ty<'tcx>,
-    ) -> EncodingResult<(vir::Expr, ty::Ty<'tcx>, Option<usize>)> {
+    ) -> EncodingResult<(PlaceEncoding, ty::Ty<'tcx>, Option<usize>)> {
```
I find a bit odd to require a `PlaceEncoding` and then immediately call `try_into_expr`. Isn't it better to keep `vir::Expr`? All calls to this function seem to be `.encode_deref(PlaceEncoding::Expr(..), ..)`.

dario23 replied:
well, it's called from encode_projection (because `Deref` is a `ProjectionElem` variant), so it seemed to make sense to be able to handle `PlaceEncoding`s there as well, so for
```
fn deref(x: [&mut u32; 2]) {
    *x[0] = 42;
}
```
we would have that combination.

on the other hand, the implementation is not there yet. i think in general there needs to be more handling in some of the places that `try_into_expr` is currently used to be able to support the full set of nesting possibilities.

on the other hand, we might go with the smaller set of changes here and leave this as-is, returning `vir::Expr`, and only change once we actually need it. that way we avoid the wrapping and unwrapping back and forth and are also clearer about the (current) actual requirements/expectations of what the method does.

fpoli replied:
I see, I missed the call from `encode_projection`. For the following case, however, wouldn't Prusti generate an internal error instead of an unsupported error because of the `try_into_expr`?
```
fn deref(x: [&mut u32; 2]) {
    *x[0] = 42;
}
```
It might be easier to keep the signature of `encode_deref` as it was and do the conversion from `PlaceEncoding` on the call site. That way, you can generate an "unsupported error" in `encode_projection`.

dario23 replied:
i'll rework this then, so that `encode_place` and `encode_projection` are the only two that return `PlaceEncoding`, and mixed/complicated cases like `encode_deref` will return `EncodingError::unsupported` for now

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5374,6 +5411,62 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+        let (encoded_expr, encoding_stmts) = self.postprocess_place_encoding(encoded_place)?;
+
+        Ok((encoded_expr, encoding_stmts, ty, variant_idx))
+    }
```
Just looking at the code in this PR, I think that this function is not needed. The only usage seems to be `self.encode_deref(PlaceEncoding::Expr(..), ..)` and `MirEncoder::encode_deref` only returns `PlaceEncoding::Expr(..)`. Do you foresee some new use cases?

dario23 replied:
removed it, as discussed in the previous comment thread.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -1287,9 +1288,12 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+
         let encode = |rhs_place| {
-            let (restored, _, _) = self.mir_encoder.encode_place(rhs_place).unwrap();
+            // TODO: pre_stmts here as well..
```
What does this comment mean?

dario23 replied:
that i should be adding an assert here as well. will do together with all the other places it's missing.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -1287,9 +1288,12 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         let encode = |rhs_place| {
-            let (restored, _, _) = self.mir_encoder.encode_place(rhs_place).unwrap();
+            // TODO: pre_stmts here as well..
+            let (restored, pre_stmts, _, _) = self.encode_place(rhs_place).unwrap();
```
Adding `assert!(pre_stmts.is_empty(), "TODO");` as you did in other places would be very good.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -3666,7 +3690,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
                 if self.mir_encoder.is_reference(arg_ty) {
                     let encoded_arg = self.mir_encoder.encode_local(arg_index)?;
                     let (deref_place, ..) =
```
This is ignoring a `pre_stms` that you could assert to be empty.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -3851,20 +3877,19 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
-            );
-            let (encoded_deref, ..) = self.mir_encoder
-                .encode_deref(encoded_return, ty)
+            let (encoded_deref, pre_stmts, ..) = self.mir_encoder
```
Assert that `pre_stmts` is empty?

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4043,7 +4068,8 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
                 // will panic if attempting to encode unsupported type
-                let (encoded_place, ty, _) = self.mir_encoder.encode_place(&mir_place).unwrap();
+                let (encoded_place, pre_stmts, ty, _) = self.encode_place(&mir_place).unwrap();
+                // TODO: pre_stmts
```
Assert that `pre_stmts` is empty?

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4074,7 +4100,8 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
                             // will panic if attempting to encode unsupported type
-                            let (_, ref_ty, _) = self.mir_encoder.encode_place(&base).unwrap();
+                            let (_, _, ref_ty, _) = self.encode_place(&base).unwrap();
+                            // TODO: pre_stmts
```
Assert that `pre_stmts` is empty?

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4135,8 +4162,9 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
-                                            self.mir_encoder.encode_place(&child_place).unwrap(); // will panic if attempting to encode unsupported type
+                                        let (encoded_child, pre_stmts, _, _) =
+                                            self.encode_place(&child_place).unwrap(); // will panic if attempting to encode unsupported type
+                                        // TODO: pre_stmts
```
Assert that `pre_stmts` is empty?

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5374,6 +5411,62 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+        Ok((encoded_expr, encoding_stmts, ty, variant_idx))
+    }
+
+    fn encode_projection(
```
Just looking at the code in this PR, it seems that this function is not needed.

dario23 replied:
`encode_projection` is called in `ProcedureEncoder::encode_statement_at`, which has a `prusti_interface::environment::mir_utils::MirPlace` instead of a `rustc_middle::mir::Place`, so we can't call `encode_place` here, nor do i see how to recover/rebuild a `mir::Place` again from that.

fpoli replied:
Isn't `ProcedureEncoder::encode_statement_at` calling `MirEncoder` with `self.mir_encoder.encode_projection`? I can't find a call of this `encode_projection`, but I might be missing something because I'm just doing textual search on the diff.

dario23 replied:
uhm.. that should be [part of my changes](https://github.com/viperproject/prusti-dev/pull/470/files#diff-704bbc6b705baff9ae5a91e768c083923396f2cfd663646531528a9494384a23R1056). didn't show up when text-searching in the diff first, because the diff in `procedure_encoder.rs` was too big so github hid it by default :-/

fpoli replied:
Yes, found it, sorry.

fpoli commented on prusti-viper/src/encoder/pure_function_encoder.rs:
```diff
@@ -491,6 +491,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> PureFunctionBackwardInterpreter<'p, 'v, 'tcx> {
                     place.local,
                     &place.projection[..],
                 ).with_span(span)?;
+                let encoded_place = encoded_place.try_into_expr().with_span(span)?;
```
`try_into_expr` is fine for now, but keep in mind that you'll need to implement for `PureFunctionEncoder` the analogous version of `ProcedureEncoder::postprocess_place_encoding` in order to encode array operations in pure functions and specifications.

dario23 replied:
hm, i guess i should do at least the same wrapping with a local version of `encode_place` and introduce a `postprocess_place_encoding` function in `PureFunctionEncoder` as well; the current way with just `try_into_expr` sprinkled everywhere doesn't seem to be clean nor meant to stay fwiw.

dario23 committed:
Move PlaceEncoding to its own file

dario23 committed:
Wrap encode_place in pure_function_encoder.rs as well

dario23 committed:
Don't return PlaceEncoding from MirEncoder::encode_deref

dario23 committed:
Always assert!(pre_stmts.is_empty()) if we can't directly handle them

This way we make sure that an error will surface and not silently ignore
statements that would have been necessary for the encoding.
